File: RiceCooker.test.ts
import { RiceCooker } from './RiceCooker';

describe('RiceCooker', () => {
  let riceCooker: RiceCooker;

  beforeEach(() => {
    riceCooker = new RiceCooker();
  });

  it('should toggle pluggedIn status when plug is called', () => {
    expect(riceCooker.pluggedIn).toBeFalsy();
    riceCooker.plug();
    expect(riceCooker.pluggedIn).toBeTruthy();
    riceCooker.plug();
    expect(riceCooker.pluggedIn).toBeFalsy();
  });

  it('should handle exceeding maxCapacity in setQuantity', () => {
    expect(riceCooker.quantity).toBe(0);
    riceCooker.setQuantity(2000);
    expect(riceCooker.quantity).toBe(2000);
    riceCooker.setQuantity(4000);
    expect(riceCooker.quantity).toBe(0)
  });

  it('should update cookTime when positive in setCookTime', () => {
    expect(riceCooker.cookTime).toBe(0);
    riceCooker.setCookTime(30);
    expect(riceCooker.cookTime).toBe(30);
    riceCooker.setCookTime(-10);
    expect(riceCooker.cookTime).toBe(0)
  });

  it('should update waterAmount when positive in setWaterAmount', () => {
    expect(riceCooker.waterAmount).toBe(0);
    riceCooker.setWaterAmount(500);
    expect(riceCooker.waterAmount).toBe(500);
    riceCooker.setWaterAmount(-50);
  });
});

File: rice_cooker/test/rice_cooker_test.dart
import 'package:rice_cooker/rice_cooker.dart';
import 'package:test/test.dart';

void main() {
  group('RiceCooker', () {
    test('plug() should toggle pluggedIn status', () {
      RiceCooker riceCooker = RiceCooker();

      expect(riceCooker.pluggedIn, isFalse);

      riceCooker.plug();
      expect(riceCooker.pluggedIn, isTrue);

      riceCooker.plug();
      expect(riceCooker.pluggedIn, isFalse);
    });

    test('setQuantity() should update quantity within maxCapacity', () {
      RiceCooker riceCooker = RiceCooker();

      expect(riceCooker.quantity, equals(0));

      riceCooker.setQuantity(2000);
      expect(riceCooker.quantity, equals(2000));

      riceCooker.setQuantity(4000);
      expect(riceCooker.quantity, equals(0));
    });

    test('setCookTime() should update cookTime when positive', () {
      RiceCooker riceCooker = RiceCooker();

      expect(riceCooker.cookTime, equals(0));

      riceCooker.setCookTime(30);
      expect(riceCooker.cookTime, equals(30));

      riceCooker.setCookTime(-10);
      expect(riceCooker.cookTime, equals(30));
    });

    test('setWaterAmount() should update waterAmount when positive', () {
      RiceCooker riceCooker = RiceCooker();

      expect(riceCooker.waterAmount, equals(0));

      riceCooker.setWaterAmount(500);
      expect(riceCooker.waterAmount, equals(500));

      riceCooker.setWaterAmount(-50);
      expect(riceCooker.waterAmount, equals(500));
    });

    test('cookRice() should print appropriate messages', () {
      RiceCooker riceCooker = RiceCooker();

      riceCooker.cookRice();

      riceCooker.plug();
      riceCooker.cookRice();

      riceCooker.setQuantity(2000);
      riceCooker.cookRice();

      riceCooker.setWaterAmount(500);
      riceCooker.cookRice();
    });
  });
}
